crypto/statetrie/nibbles/nibbles_test.go

```diff
@@ -0,0 +1,154 @@
+		{0x4, 0x1, 0x2, 0x3, 0x4},
+		{0x0, 0x0, 0x2, 0x3, 0x5},
+		{0x0, 0x1, 0x2, 0x3, 0x4, 0x5},
+		{},
+		{0x1},
+	}
+
+	sampleNibblesPacked := [][]byte{
+		{0x01, 0x23, 0x40},
+		{0x41, 0x23, 0x40},
+		{0x00, 0x23, 0x50},
+		{0x01, 0x23, 0x45},
+		{},
+		{0x10},
+	}
+
+	sampleNibblesShifted1 := []Nibbles{
+		{0x1, 0x2, 0x3, 0x4},
+		{0x1, 0x2, 0x3, 0x4},
+		{0x0, 0x2, 0x3, 0x5},
+		{0x1, 0x2, 0x3, 0x4, 0x5},
+		{},
+		{},
+	}
+
+	sampleNibblesShifted2 := []Nibbles{
+		{0x2, 0x3, 0x4},
+		{0x2, 0x3, 0x4},
+		{0x2, 0x3, 0x5},
+		{0x2, 0x3, 0x4, 0x5},
+		{},
+		{},
+	}
+
+	for i, n := range sampleNibbles {
+		b, oddLength := Pack(n)
+		if oddLength {
+			// require that oddLength packs returns a byte slice with the last nibble set to 0x0
+			require.Equal(t, b[len(b)-1]&0x0f == 0x00, true)
+		}
+
+		require.Equal(t, oddLength == (len(n)%2 == 1), true)
+		require.Equal(t, bytes.Equal(b, sampleNibblesPacked[i]), true)
+
+		unp := Unpack(b, oddLength)
+		require.Equal(t, bytes.Equal(unp, n), true)
+
+	}
+	for i, n := range sampleNibbles {
+		require.Equal(t, bytes.Equal(ShiftLeft(n, -2), sampleNibbles[i]), true)
+		require.Equal(t, bytes.Equal(ShiftLeft(n, -1), sampleNibbles[i]), true)
+		require.Equal(t, bytes.Equal(ShiftLeft(n, 0), sampleNibbles[i]), true)
+		require.Equal(t, bytes.Equal(ShiftLeft(n, 1), sampleNibblesShifted1[i]), true)
+		require.Equal(t, bytes.Equal(ShiftLeft(n, 2), sampleNibblesShifted2[i]), true)
+	}
+
+	sampleSharedNibbles := [][]Nibbles{
+		{{0x0, 0x1, 0x2, 0x9, 0x2}, {0x0, 0x1, 0x2}},
+		{{0x4, 0x1}, {0x4, 0x1}},
+		{{0x9, 0x2, 0x3}, {}},
+		{{0x0}, {0x0}},
+		{{}, {}},
+	}
+	for i, n := range sampleSharedNibbles {
+		shared := SharedPrefix(n[0], sampleNibbles[i])
+		require.Equal(t, bytes.Equal(shared, n[1]), true)
+		shared = SharedPrefix(sampleNibbles[i], n[0])
+		require.Equal(t, bytes.Equal(shared, n[1]), true)
+	}
+
+	sampleSerialization := []Nibbles{
+		{0x0, 0x1, 0x2, 0x9, 0x2},
+		{0x4, 0x1},
+		{0x4, 0x1, 0x4, 0xf},
+		{0x4, 0x1, 0x4, 0xf, 0x0},
+		{0x9, 0x2, 0x3},
+		{},
+		{0x05},
+		{},
+	}
+
+	for _, n := range sampleSerialization {
+		nbytes := Serialize(n)
+		n2, err := DeserializeNibbles(nbytes)
+		require.NoError(t, err)
+		require.Equal(t, bytes.Equal(n, n2), true)
+	}
+
+	makeNibblesTestExpected := Nibbles{0x0, 0x1, 0x2, 0x9, 0x2}
+	makeNibblesTestData := []byte{0x01, 0x29, 0x20}
+	mntr := MakeNibbles(makeNibblesTestData, true)
+	require.Equal(t, bytes.Equal(mntr, makeNibblesTestExpected), true)
+	makeNibblesTestExpectedFW := Nibbles{0x0, 0x1, 0x2, 0x9, 0x2, 0x0}
+	mntr2 := MakeNibbles(makeNibblesTestData, false)
+	require.Equal(t, bytes.Equal(mntr2, makeNibblesTestExpectedFW), true)
+
+	sampleEqualFalse := [][]Nibbles{
+		{{0x0, 0x1, 0x2, 0x9, 0x2}, {0x0, 0x1, 0x2, 0x9}},
+		{{0x0, 0x1, 0x2, 0x9}, {0x0, 0x1, 0x2, 0x9, 0x2}},
+		{{0x0, 0x1, 0x2, 0x9, 0x2}, {}},
+		{{}, {0x0, 0x1, 0x2, 0x9, 0x2}},
+		{{0x0}, {}},
+		{{}, {0x0}},
+		{{}, {0x1}},
+	}
+	for _, n := range sampleEqualFalse {
```

Comment:
add a random test case for even and odd length for Serialize/ Deserialize and Pack/Unpack

Reply:
Added 1,000 generated nibbles from random data, sometimes odd sometimes even, and sometimes with the last nibble cleared in odd ones, then ran operations using the four methods [#a33795](https://github.com/algorand/go-algorand/pull/5719/commits/a33795cd11c5ca838a5d6b84e9f30a7ad2d8e320)